Source/WebKit/WebProcess/WebPage/WebFrame.cpp

```diff
@@ -1318,4 +1318,36 @@ bool WebFrame::isFocused() const
     return m_coreFrame->page()->focusController().focusedFrame() == coreFrame();
 }
 
+String WebFrame::frameTextForTesting(bool includeSubframes)
+{
+    if (!m_coreFrame)
+        return { };
+
+    StringBuilder builder;
+
+    // To keep things tidy, strip all trailing spaces: they are not a meaningful part of dumpAsText test output.
+    // Breaking the string up into lines lets us efficiently strip and has a side effect of adding a newline after the last line.
+    if (String text = innerText(); !!text) {
+        for (auto line : StringView(text).splitAllowingEmptyEntries('\n')) {
+            while (line.endsWith(' '))
+                line = line.left(line.length() - 1);
+            builder.append(line, '\n');
+        }
+    }
+
+    if (!includeSubframes)
+        return builder.toString();
+
+    for (auto* child = m_coreFrame->tree().firstChild(); child; child = child->tree().nextSibling()) {
+        if (is<RemoteFrame>(*child)) {
+            builder.append(m_page->sendSync(Messages::WebPageProxy::FrameTextForTesting(child->frameID())).takeReplyOr("Sending WebPageProxy::FrameTextForTesting failed"_s));
+        } else if (RefPtr childWebFrame = fromCoreFrame(*child)) {
```

Comment:
Similar here, checking `!childWebFrame->innerText().isNull()` in this if statement would match previous behavior.